build/gen_stub.php

```diff
@@ -1591,24 +1603,56 @@ function (Expr $expr) use ($allConstInfos, &$constType, &$originatingConst, &$is
                 }
 
                 if ($originatingConstName->isUnknown()) {
-                    $originatingConst = null;
-                    $isUnknownConstValue = true;
+                    return;
+                }
+
+                foreach ($this->allConstInfos as $const) {
+                    if ($originatingConstName->equals($const->name)) {
+                        $this->visitedConstants[] = $const;
+                        return $const->getValue($this->allConstInfos)->expr;
+                    }
+                }
+            }
+        };
+
+        $nodeTraverser = new PhpParser\NodeTraverser;
+        $nodeTraverser->addVisitor($visitor);
+        $expr = $nodeTraverser->traverse([$expr])[0];
+
+        $isUnknownConstValue = false;
 
+        $evaluator = new ConstExprEvaluator(
+            function (Expr $expr) use ($allConstInfos, &$isUnknownConstValue) {
+                if (!$expr instanceof Expr\ConstFetch) {
+                    throw new Exception($this->getVariableTypeName() . " " . $this->getVariableLikeName() . " has an unsupported value");
+                }
+
+                $constName = $expr->name->__toString();
+                if (strtolower($constName) === "unknown") {
+                    $isUnknownConstValue = true;
                     return null;
                 }
 
                 foreach ($allConstInfos as $const) {
-                    if (!$originatingConstName->equals($const->name)) {
+                    if ($constName != $const->cValue) {
                         continue;
                     }
 
-                    if ($constType === null && $const->phpDocType) {
+                    if ($const->phpDocType) {
                         $constType = $const->phpDocType->tryToSimpleType();
+                        if ($constType->isBool()) {
+                            return true;
+                        } elseif ($constType->isInt()) {
+                            return 1;
+                        } elseif ($constType->isFloat()) {
+                            return M_PI;
```

Comment:
just some random float which doesn't end up looking like an int. The purpose here is doing the operation with values of that type, so that we know the actual type after evaluation.
Pretty primitive, probably not perfect, but can be improved when needed.